components/brave_shields/browser/brave_shields_util.cc

```diff
@@ -335,50 +406,121 @@ DomainBlockingType GetDomainBlockingType(HostContentSettingsMap* map,
+                                      CONTENT_SETTING_ALLOW);
+        profile_state->SetInteger(
+            ::prefs::kCookieControlsMode,
+            static_cast<int>(
+                content_settings::CookieControlsMode::kBlockThirdParty));
+        break;
+      default:
+        NOTREACHED() << "Invalid ControlType for cookies";
+    }
+    return;
+  }
+
+  map->SetContentSettingCustomScope(host_pattern,
                                     ContentSettingsPattern::Wildcard(),
                                     ContentSettingsType::BRAVE_REFERRERS,
                                     GetDefaultBlockFromControlType(type));
+  const auto primary_pattern = CreatePrimaryPattern(host_pattern);
 
-  map->SetContentSettingCustomScope(
-      primary_pattern,
-      ContentSettingsPattern::FromString("https://firstParty/*"),
-      ContentSettingsType::BRAVE_COOKIES, GetDefaultAllowFromControlType(type));
-
-  map->SetContentSettingCustomScope(
-      primary_pattern, ContentSettingsPattern::Wildcard(),
-      ContentSettingsType::BRAVE_COOKIES, GetDefaultBlockFromControlType(type));
-
-  RecordShieldsSettingChanged(local_state);
+  switch (type) {
+    case ControlType::BLOCK_THIRD_PARTY:
+      // general-rule:
+      map->SetContentSettingCustomScope(
+          ContentSettingsPattern::Wildcard(), host_pattern,
+          ContentSettingsType::BRAVE_COOKIES, CONTENT_SETTING_BLOCK);
+      // first-party rule:
+      map->SetContentSettingCustomScope(primary_pattern, host_pattern,
+                                        ContentSettingsType::BRAVE_COOKIES,
+                                        CONTENT_SETTING_ALLOW);
+      break;
+    case ControlType::ALLOW:
+    case ControlType::BLOCK:
+      // Remove first-party rule:
+      map->SetContentSettingCustomScope(primary_pattern, host_pattern,
+                                        ContentSettingsType::BRAVE_COOKIES,
+                                        CONTENT_SETTING_DEFAULT);
+      // general-rule:
+      map->SetContentSettingCustomScope(
+          ContentSettingsPattern::Wildcard(), host_pattern,
+          ContentSettingsType::BRAVE_COOKIES,
+          (type == ControlType::ALLOW) ? CONTENT_SETTING_ALLOW
+                                       : CONTENT_SETTING_BLOCK);
+      break;
+    default:
+      NOTREACHED() << "Invalid ControlType for cookies";
+  }
 }
 
-ControlType GetCookieControlType(HostContentSettingsMap* map, const GURL& url) {
-  ContentSetting setting =
-      map->GetContentSetting(url, GURL(), ContentSettingsType::BRAVE_COOKIES);
+ControlType GetCookieControlType(
+    HostContentSettingsMap* map,
+    content_settings::CookieSettings* cookie_settings,
+    const GURL& url) {
+  DCHECK(map);
+  DCHECK(cookie_settings);
 
-  ContentSetting fp_setting = map->GetContentSetting(
-      url, GURL("https://firstParty/"), ContentSettingsType::BRAVE_COOKIES);
+  auto result = CookieRules::Get(map, url, ContentSettingsType::COOKIES);
```

Comment:
This should be BRAVE_COOKIES because otherwise the control will show the combined value of shield and setting cookies (incorrectly in many cases) and not be an accurate reflection of the actual setting. See https://github.com/brave/brave-browser/issues/24321